contracts/LoanCore.sol

```diff
@@ -743,12 +800,20 @@ contract LoanCore is
         if (protocolFee > 0) _feesWithdrawable[address(this)] += protocolFee;
         if (affiliateFee > 0) _feesWithdrawable[affiliate] += affiliateFee;
 
-        // State changes and cleanup
-        loans[loanId].state = LoanLibrary.LoanState.Repaid;
-        loans[loanId].interestAmountPaid = _amountFromPayer - data.balance;
-        loans[loanId].balance = 0;
+        // state changes
+        loanRepaid = false;
+        if (_paymentToPrincipal >= data.balance) {
+            // if the payment is greater than or equal to the balance, the loan is repaid
+            loans[loanId].state = LoanLibrary.LoanState.Repaid;
+            // tell the calling function that the loan is repaid
+            loanRepaid = true;
+            // mark collateral as no longer escrowed
+            collateralInUse[keccak256(abi.encode(data.terms.collateralAddress, data.terms.collateralId))] = false;
+        }
+
+        loans[loanId].interestAmountPaid += _interestAmount;
+        loans[loanId].balance -= _paymentToPrincipal;
```

Comment:
Revert check for possible underflow here? From LoanCore's perspective, `_paymentToPrincipal` is untrusted input and could be any value, plus it seems like underflow is very possible (since we are handling the "**greater than** or equal to" scenario on line 805 above)

Reply:
Yes, no checked from LC's perspective. I added this line:
` if (_paymentToPrincipal > data.balance) revert LC_ExceedsBalance(_paymentToPrincipal, data.balance);`
and change the >= to be == in the check for if this is the final repayment.